swap/src/protocol/alice/state.rs

```diff
@@ -403,6 +404,55 @@ impl State3 {
         )
     }
 
+    pub async fn check_for_tx_cancel(
```

Comment:
Happy to leave them for now to move forward but overall, there isn't really any need why this functions need to sit on `State3`. You are getting as much data / functionality off `self` than you do from `bitcoin_wallet`. Putting it on `bitcoin::Wallet` is equally strange, hence I think this could just be a helper function inside the `swap` module.

Reply:
Not a recommendation to do, just want to hear your feedback: 

This function is duplicated in `state3` for Alice and `state6` for Bob. 
How about converting this function into a free-floating one which takes as input a wallet and something that implements `.get_cancel()`, that way you can remove one and it can live somewhere else.

Reply:
I would clean this stuff up in a follow up because I think a complete solution should get some more thought. I think we should find a better solution for all these function on both sides. 
If it allows re-using functionality on both sides - even better :)

Reply:
> How about converting this function into a free-floating one which takes as input a wallet and something that implements .get_cancel(), that way you can remove one and it can live somewhere else.

Abstracting over `get_cancel` would require a trait and the only benefit you are getting from that is that you can type:

```rust
check_for_tx_cancel(state, wallet)
```

instead of 


```rust
check_for_tx_cancel(state.tx_cancel(), wallet)
```

I would say it is probably not worth it to introduce a trait and having to do at least 2 implementations of it (I don't know how many different states we need to call this function from).

As a rule of thumb, abstracting over functionality with a trait is particularly useful if conditionals are involved. Let's say for example your function looks like this:

```rust
fn foo<T: Bar>(bar: &T, baz: Baz) -> u64 {
	let a_or_b = if baz.some_condition() {
		bar.compute_a()
	} else {
		bar.compute_b()
	};

	baz.compute_something(a_or_b)
}
```

In this case, having the trait is particularly useful because whether or not we want to call `compute_a` or `compute_b` depends on `Baz`. Hence, we cannot just "lift" this out and make it the caller's responsibility because we would have to lift the whole `if` block out which kind of makes the whole function pointless.

In our case, we always take `TxCancel` and pass it to the wallet, hence we can just have the caller do that and therefore avoid having to introduce a trait and several implementations at virtually no benefit.

It is usually good to strive for passing around _values_ (like `TxCancel`) and only use traits for real functionality that is not just a getter.